### DISCUSSION TITLE
ArgoCD API Patch Resource

### DISCUSSION THREAD

**Discussion by danielloynaz** [argoproj/argo-cd#9390] [Q&A] [answered]
Hi,

I'm trying to patch an application resource following the swagger documentation for the API. (https://cd.apps.argoproj.io/swagger-ui#operation/ApplicationService_PatchResource). But I'm getting error response. I have try different ways without luck.

**Request Info**
URL: https://xx.xxxxxx.xxx/api/v1/applications/<app-name>/resource

Query Params:
namespace:       my-namespace
resourceName: config-map
kind:                  ConfigMap
patchType:        application/json-patch+json

Body: (JSON)
```
{
"data": {
        ".env": ""
    }
}
```
I have tried many different formats, including, passing the parameters in the request body.
```
{
    "Namespace": "my-namespace",
    "ResourceName": "config-map",
    "Kind": "ConfigMap",
    "PatchType":"application/json-patch+json",
    "Patch": "{\"data\": {\".env\": \"\"}}"
}
```

Always getting this error response:

```
{
    "error": "json: cannot unmarshal object into Go value of type string",
    "code": 3,
    "message": "json: cannot unmarshal object into Go value of type string"
}
```
Is there a place where I can find more detailed information about the request format?
Can you provide a full example of how to use this endpoint?

Best regards,

**Comment by danielloynaz** [comment]
I'm able to patch the resources using the CLI.

```
argocd app patch-resource argocd-apps-backend --kind ConfigMap --namespace app --resource-name backend-config-map --patch-type "application/merge-patch+json" --patch '{ \"data\":{ \"config\": \"test\" } }'
```
Result:
```
time="2022-05-13T18:28:48-04:00" level=info msg="Resource 'backend-config-map' patched"
```

My API Call look like this:
```
curl --location --request POST 'https://localhost:3000/api/v1/applications/argocd-apps-backend/resource?namespace=app&resourceName=backend-config-map&kind=ConfigMap&patchType=application/merge-patch+json' \
--header 'Authorization: Bearer <TOKEN>' \
--header 'Content-Type: text/plain' \
--data-raw '{
    "data": {
        "config": "test"
    }
}'
```
Response:
```
{
    "error": "json: cannot unmarshal object into Go value of type string",
    "code": 3,
    "message": "json: cannot unmarshal object into Go value of type string"
}
```
I have tested different data formats like: `--data-raw ''\''{ \"data\":{ \"config\": \"test\" } }'\'''` but then this is the error.
```
{
    "error": "invalid character '\\'' looking for beginning of value",
    "code": 3,
    "message": "invalid character '\\'' looking for beginning of value"
}
```
Any suggestion?

**Comment by danielloynaz** [comment, marked as answer]
Problem solved!

The issue was related with the query param `patchType`. 

I was using the following value `application/merge-patch+json` but it require the value in url encoded format `application%2Fmerge-patch%2Bjson`

**Reply by fedeostrit** [reply]
Hello Daniel, good morning, I'm having the same problem when trying to make a request to the "PatchResource" method respecting the swaggerUI definitions of argoscd, the error is the same "{
    "error": "json: cannot unmarshal object into Go value of type string",
    "code": 3,
    "message": "json: cannot unmarshal object into Go value of type string"
}" I am using Postman to make the request, could you give me the same definition that you did and that worked for you, if possible, tell me how you send it by postman

**Reply by fedeostrit** [reply]
hello everyone, could you help us with this problem using the PatchResource method both by API and by ArgoCD Cli which most of the time generates the error:

{
    "error": "json: cannot unmarshal array into Go value of type string",
    "code": 3,
    "message": "json: cannot unmarshal array into Go value of type string"
}

An example of the per-cli command I'm sending:

argocd app patch-resource 'dev-devops-pepe' --group 'apps' --kind 'Deployment' --namespace 'devops-pepe' --resource-name 'devops-pepe' --patch '{\"requests\":{\"memory\":\"512M\"}}' --patch-type 'application/merge-patch+json' --loglevel 'debug' --logformat 'json'

answer: argocd : {"level":"info","msg":"Resource 'devops-pepe' patched","time":"2023-03-22T18:52:10-03:00"} but when I go to the Argos UI console I don't see the updated deployment, 
This is the live manifest that is seen on the console and I need to update {resource: {request: memory: 512mb} for another value

`apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: '7'
  creationTimestamp: '2022-06-03T22:45:28Z'
  generation: 7
  labels:
    agentName: dev
    app.kubernetes.io/instance: dev-devops-pepe
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: dev-dev-devops-pepe
    app.kubernetes.io/version: latest
    argocd.argoproj.io/instance: dev-devops-pepe
    enviroment: dev
    familyName: devops
    helm.sh/chart: microservice-0.1.0
    serviceName: pepe
  name: devops-pepe
  namespace: devops-pepe
  resourceVersion: '851219754'
  uid: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 1
  selector:
    matchLabels:
      app.kubernetes.io/instance: dev-devops-pepe
      app.kubernetes.io/name: dev-dev-devops-pepe
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      annotations:
        kubectl.kubernetes.io/restartedAt: '2022-06-03T23:14:22Z'
        sidecar.istio.io/proxyMemory: 512M
        sidecar.istio.io/proxyMemoryLimit: 2Gi
        traffic.sidecar.istio.io/excludeOutboundPorts: '8200'
        vault.hashicorp.com/role: devops-pepe
      creationTimestamp: null
      labels:
        app.kubernetes.io/instance: dev-devops-pepe
        app.kubernetes.io/name: dev-dev-devops-pepe
        family: devops
        type: microservice
    spec:
      containers:
        - env:
            - name: MS_NAME
              value: devops-pepe
            - name: FAMILYNAME
              value: devops
            - name: SERVICENAME
              value: pepe
            - name: OTEL_RESOURCE_ATTRIBUTES
              value: >-
                environment=dev,service.name=devops-pepe,service.namespace=devops,service.version=latest
          envFrom:
            - configMapRef:
                name: global-devops-pepe
            - secretRef:
                name: devops-pepe-global-kafka
            - secretRef:
                name: devops-pepe-global-redis
            - secretRef:
                name: devops-pepe-global-kafka
            - secretRef:
                name: devops-pepe-signalfx
          image: 'xxxxxxxxxxxxxxxxxxxxxxxxxxxx'
          imagePullPolicy: Always
          livenessProbe:
            failureThreshold: 3
            httpGet:
              path: /
              port: http
              scheme: HTTP
            initialDelaySeconds: 60
            periodSeconds: 5
            successThreshold: 1
            timeoutSeconds: 15
          name: dev-dev-devops-pepe
          ports:
            - containerPort: 80
              name: http
              protocol: TCP
          readinessProbe:
            failureThreshold: 3
            httpGet:
              path: /
              port: http
              scheme: HTTP
            initialDelaySeconds: 15
            periodSeconds: 5
            successThreshold: 1
            timeoutSeconds: 15
          resources:
            limits:
              memory: 1G
            requests:
              memory: 512M
          securityContext: {}
          startupProbe:
            failureThreshold: 10
            httpGet:
              path: /
              port: http
              scheme: HTTP
            periodSeconds: 30
            successThreshold: 1
            timeoutSeconds: 1
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      nodeSelector:
        node_group: microservices
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: devops-pepe
      serviceAccountName: devops-pepe
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
    - lastTransitionTime: '2023-03-02T09:49:51Z'
      lastUpdateTime: '2023-03-02T09:49:51Z'
      message: Deployment has minimum availability.
      reason: MinimumReplicasAvailable
      status: 'True'
      type: Available
    - lastTransitionTime: '2022-06-03T23:14:22Z'
      lastUpdateTime: '2023-03-22T21:24:48Z'
      message: ReplicaSet "devops-pepe-994b868d" has successfully progressed.
      reason: NewReplicaSetAvailable
      status: 'True'
      type: Progressing
  observedGeneration: 7
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
`

@crenshaw-dev , @jannfis , @jessesuen, @alexmt, @alexec 

Could you help me to validate why the change is not seen when I send it by argocd cli and why if I send a POST request to the PatchResource api sending the body "{\"requests\":{\"memory\":\"512M\" }}" returns "json: cannot unmarshal array into Go value of type string"

**Reply by modernfaust** [reply]
This solution doesn't work for me. I'm sending

`curl -X POST -H 'Content-Type: application/json' -H 'Accept: application/json' -H "Authorization: Bearer ${ARGO_JWT_TOKEN}" -k https://${ARGO_SERVER_URL}/api/v1/applications/${ARGO_APP_NAME}/resource?version=${ARGO_APPSET_VERSION}\&patchType=application%2Fmerge-patch%2Bjson --data '{ "spec": { "spec": { "syncPolicy": "{}"}}}'
`

and receiving

`{"error":"json: cannot unmarshal object into Go value of type string","code":3,"message":"json: cannot unmarshal object into Go value of type string"}bash-3.2$ `

**Comment by jwenz723** [comment]
The solution mentioned did not work for me. I was doing raw HTTP POST requests similar to you. I found I was able to resolve the problem by migrating to making use of the ArgoCD client provided by the ArgoCD github repo. I believe the client is making gRPC calls rather than typical REST HTTP calls.

Here is an example of how to create an ArgoCD client and invoke the PatchResource APi.

```golang
package main

import (
	"context"
	"fmt"
	"github.com/argoproj/argo-cd/v2/pkg/apiclient"
	"github.com/argoproj/argo-cd/v2/pkg/apiclient/application"
	"os"
)

func main() {
	token := os.Getenv("ARGOCD_TOKEN")
	argoCDURL := os.Getenv("ARGOCD_URL")

	apiClient, err := apiclient.NewClient(&apiclient.ClientOptions{
		ServerAddr: fmt.Sprintf(argoCDURL),
		GRPCWeb:    true,
		AuthToken:  token,
	})
	if err != nil {
		panic(fmt.Errorf("failed to create argocd api client: %w", err))
	}

	_, appClient, err := apiClient.NewApplicationClient()
	if err != nil {
		panic(fmt.Errorf("failed to create argocd application client: %w", err))
	}

	_, err = appClient.PatchResource(context.TODO(), &application.ApplicationResourcePatchRequest{
		// request data here
	})
	if err != nil {
		panic(err)
	}
}

```

**Comment by DenisMedeiros** [comment]
Also had a similar issue. The solution: edited the live manifest using Google Chome and copied the request from the Networking tab to see the right structure and parameters.

In the example below, we were able to modify an ApplicationSet that is created by an Application (that is, the resource is an ApplicationSet but the same idea can be used to patch any resource that is a child of an Application):

```bash
curl 'https://<your-argo-server>/api/v1/applications/<app-name>/resource?name=<appset-name>&appNamespace=<app-namespace-in-argocd>&namespace=<appset-namespace>&resourceName=<appset-name>&version=v1alpha1&kind=ApplicationSet&group=argoproj.io&patchType=application%2Fmerge-patch%2Bjson' \
  --header 'Content-Type: application/json' \
  --header 'Authorization: Bearer <token>' \
  --data-raw '"{\"spec\":{\"template\":{\"spec\":{\"syncPolicy\":{\"automated\":{\"prune\":true,\"selfHeal\":true},\"managedNamespaceMetadata\":{\"labels\":{\"argocd.argoproj.io/managed-by\":\"openshift-gitops\"}},\"retry\":{\"backoff\":{\"duration\":\"15s\",\"factor\":2,\"maxDuration\":\"15m\"},\"limit\":7},\"syncOptions\":[\"CreateNamespace=true\"]}}}}}"'
```

Another example modifying a service (adding/ modifying an annotation) that is part of an application:

```bash
curl '<your-argo-server>/api/v1/applications/<app-name>/resource?name=<service-nakme>&appNamespace=<app-namespace-in-argocd>&namespace=<service-namespace>&resourceName=<service-name>&version=v1&kind=Service&group=&patchType=application%2Fmerge-patch%2Bjson' \
  -H 'Content-Type: application/json' \
  -H 'Authorization: Bearer <token>' \
  -X 'POST' \
  --data-raw '"{\"metadata\":{\"labels\":{\"test\":\"123\"}}}"'
```
